plugin/cog/cog-rangereader-azure/src/main/java/it/geosolutions/imageioimpl/plugins/cog/AzureClientFactory.java

```diff
@@ -0,0 +1,54 @@
+/*
+ *    ImageI/O-Ext - OpenSource Java Image translation Library
+ *    http://www.geo-solutions.it/
+ *    https://github.com/geosolutions-it/imageio-ext
+ *    (C) 2022, GeoSolutions
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    either version 3 of the License, or (at your option) any later version.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package it.geosolutions.imageioimpl.plugins.cog;
+
+import it.geosolutions.imageio.core.ExtCaches;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utility class to assist building Azure async client.
+ * Azure clients should be singletons and re-used.
+ */
+public class AzureClientFactory {
+
+    private AzureClientFactory() {
+    }
+
+    private static Map<String, AzureClient> asyncClients = new HashMap<>();
+
+    public static AzureClient getClient(AzureConfigurationProperties configProps) {
+
+        String container = configProps.getContainer();
+        if (asyncClients.containsKey(container)) {
```

Comment:
The S3 client factory also has a cache, but it's based on regions, which are maybe a dozen. In Azure the container seems to be something similar to a S3 bucket, so there might be many more.. wondering if this might grow to the point of being a problem.... likely not, unless a server ends up talking with many of them... but maybe worth considering the SoftValueHash map available in imageio-ext-utilities.

Reply:
Thanks for the suggestion. I'll take a look on it.

Reply:
Fixed